src/utilities/contextual-code.ts

```diff
@@ -0,0 +1,168 @@
+
+import path = require('path');
+import * as vscode from 'vscode';
+import { SemanticTokensRegistrationType, SemanticTokensProviderShape, SymbolKind } from 'vscode-languageclient';
+import { ILspAnalyzer } from '../shared/types/LspAnalyzer';
+import { Outline } from '../shared/types/custom_protocols';
+import { getCodeForRange, isPositionInElementDefinitionRange } from '../shared/utils';
+import { Token } from '../shared/types/token';
+export class ContextualCodeProvider {
+
+    public async getContextualCode(document: vscode.TextDocument, range: vscode.Range, analyzer: ILspAnalyzer): Promise<string | undefined> {
+        const checkSymbols = (symbols: Outline[]): Outline | undefined => {
+            for (const symbol of symbols) {
+                const symbolRange = new vscode.Range(
+                    symbol.range.start.line,
+                    symbol.range.start.character,
+                    symbol.range.end.line,
+                    symbol.range.end.character,
+                );
+                if (isPositionInElementDefinitionRange(symbol, range.start)) {
+                    return symbol;
+                }
+                if (symbol.children) {
+                    const result = checkSymbols(symbol.children);
+                    if (result !== undefined) {
+                        return result;
+                    }
+                }
+            }
+            return undefined;
+        };
+        const outline = await analyzer.fileTracker.waitForOutline(document);
+
+        if (outline === undefined) {
+            return undefined;
+        }
+        const outlineSymbols = outline?.children || [];
+        // parse the outline to get all top level symbols
+        const symbol = checkSymbols(outlineSymbols);
```

Comment:
Please document the requirement of this check and recheck the code once. there seems to be some issues.